### ISSUE TITLE
Rename is not updating all usages.

### ISSUE THREAD

**Issue by as3boyan** [TiVo/intellij-haxe#222]
```haxe
package foo;
import foo.MainClass; //Doesn't make any sense - and it should be marked as unused
class MainClass {
  var main:Main<caret>Class;
}
```
Class needs to be imported elsewhere. Or just rename without import first.

Anyway, even in this case, plugin should understand that it's the same class, and rename all references.

About Travis CI: it's a great and very interesting opportunity, but currently it spams me with failing cases on each commit in pull request, so I have to disable email notifications.

**Comment by as3boyan** [comment]
```haxe
package a;
import a.MainClass;
class MainClass { //Find usages won't find this reference, should it? Need to look in Java
    var main:MainClass;
}
```

**Comment by as3boyan** [comment]
Ok, so, the things is, rename renames class and file, but when it proceeds to usages, it checks if references can be resolved, after renaming class, they are no longer valid anymore of course

**Comment by EBatTiVo** [comment]
So we have to collect all of the references before you allow the rename to proceed.  Right?  Don't we have to do this anyway, in order to put up the dialog?

**Comment by as3boyan** [comment]
I'm not sure now, I don't get it yet, but it seems like HaxeReferenceExpression, has bindToElement, which makes reference to point to file/package, not sure about class

**Comment by as3boyan** [comment]
I think Move File functionality may be similar to this, but still I don't really understand how in old version this compiles and works fine, but in new we have to implement it.

**Comment by EBatTiVo** [comment]
My solution is to check out an old version (eb94f32bbaccf3b3b193303ad340c00e337d5f2f is just before we started the ClassHierarchy work), watch it work, then git bisect until I find the checkin that broke it.  Takes about 12 builds and I usually have the change in 20 minutes.  Then, I inspect the change (git bisect visualize), check out the immediately previous (working) code and make little independent changes until I figure out which one broke things.

If git bisect doesn't work for you, then you might have to do some more sleuthing to see if it was ever implemented.

**Comment by as3boyan** [comment]
rename/move works here: https://github.com/TiVo/intellij-haxe/commit/8f9adcbad4ee4b8761b9a963c130ede2d75702f8

next commit - doesn't works: https://github.com/TiVo/intellij-haxe/commit/18d936702a7f93b9df0b244a4101ed2f3df669e2

**Comment by sganapavarapu1** [comment]
was reviewing Move functioning on Friday, found that - 
> before move: findUsages part of Move was returning correct list of usages (be it import statements, or variable type declarations, or someVar.getPropertyOfTypeOfClassBeingMoved() etc.)
> after move: I would expect it to correctly fix all those instances from findUsages list

unfortunately, it will doing everything but 'import' statement occurrences. 

There wasn't any implementation for Move earlier - it was using underlying common implementation of open api. But now, there's a need to fix this!

@as3boyan When I look at the diff of the commits you mentioned above... it is not clear how rename/move broke between those 2 commits :(

I'll try to figure out either how Move broke? or implement the Move file/package functionality over next 2-3 days.

**Comment by as3boyan** [comment]
I was testing commits before these, this one may be breaking https://github.com/TiVo/intellij-haxe/commit/be548c1696f0d3379d266a501cd4e4e611965fce

It seems like this commit broke `testMovePackage`

If we could build every commit and see when something got broken, that would be awesome.

About `git bisect` I need to try it sounds good.

**Comment by as3boyan** [comment]
I got `testMovePackage` to work and almost implemented some rename/move, but not sure how it worked before.

@sganapavarapu1 

>There wasn't any implementation for Move earlier - it was using underlying common implementation of open api. But now, there's a need to fix this!

I understand, but I would prefer that common implementation.

**Comment by as3boyan** [comment]
>unfortunately, it was only fixing part of findUsages returned list... it wasn't fixing "import" statement occurrences. 

So previous (common) implementation didn't worked this case?

**Comment by sganapavarapu1** [comment]
@as3boyan 
yes, I too prefer common implementation
I'm not sure whether previous implementation has 'import' statements also working!

Today, I was about to start working on `testMovePackage`, `testMoveFile1`, `testMoveFile2` - but I will not do that now because you already fixed `testMovePackage` and you are working on some rename/move. It will be duplication of effort - I will pick up something else.

Thanks,

**Comment by as3boyan** [comment]
https://github.com/TiVo/intellij-haxe/commit/4b822108dc322edc122dfe06aa34e37b5b081253 may break `testMoveFile2`

**Comment by EBatTiVo** [comment]
@as3boyan I see a PR was merged for this.  Is this fixed?

**Comment by as3boyan** [comment]
@EBatTiVo It was for `testMovePackage` I think

**Comment by as3boyan** [comment]
https://github.com/TiVo/intellij-haxe/commit/858ef3adfd04a3bfb6cc8ec9e08e78b3b8e06cc6 is indeed broke the move file and package because
`HaxeTypeTag typeTag = PsiTreeUtil.getChildOfType(parent, HaxeTypeTag.class);` used to throw `unimplemented` error for PsiPackage something like that

**Comment by as3boyan** [comment]
But it's different error, and if I revert change in master branch - that doesn't fixes the issue.

**Comment by as3boyan** [comment]
It has to do something with resolving/checking each usage using `isReferenceTo`

**Comment by as3boyan** [comment]
@EBatTiVo indeed PR was for rename, but not sure about it, from some point it may be enough to get tests to work, it might be connected to move functionality. Need to investigate.

Next `git bisect` bad commit is e1822d7439fd79be6f4b1b5ec6ae3c06e1862052.
Is more likely to break move. Will look into it.

**Comment by as3boyan** [comment]
I had to revert changes in https://github.com/TiVo/intellij-haxe/commit/e1822d7439fd79be6f4b1b5ec6ae3c06e1862052, to make it work

@EBatTiVo  Please look into https://github.com/TiVo/intellij-haxe/pull/271

**Comment by as3boyan** [comment]
#272 should fix this issue

**Comment by EBatTiVo** [comment]
Fixed with #272.  Closing.
